rust/src/parse.rs

```diff
@@ -0,0 +1,392 @@
+    let (name, data) = parse_str(data)?;
+    let (metadata, _) = parse_str_map(data)?;
+    Ok(Metadata {
+        name: name.into(),
+        metadata,
+    })
+}
+
+fn parse_metadata_index(data: &[u8]) -> Result<MetadataIndex<'_>, ParseError> {
+    let (offset, data) = parse_u64(data)?;
+    let (length, data) = parse_u64(data)?;
+    let (name, _) = parse_str(data)?;
+    Ok(MetadataIndex {
+        offset,
+        length,
+        name: name.into(),
+    })
+}
+
+fn parse_summary_offset(data: &[u8]) -> Result<SummaryOffset, ParseError> {
+    if data.is_empty() {
+        return Err(ParseError::DataTooShort);
+    }
+    let (opcode_buf, data) = data.split_at(1);
+    let group_opcode = opcode_buf[0];
+    let (group_start, data) = parse_u64(data)?;
+    let (group_length, _) = parse_u64(data)?;
+    Ok(SummaryOffset {
+        group_opcode,
+        group_start,
+        group_length,
+    })
+}
+
+fn parse_data_end(data: &[u8]) -> Result<DataEnd, ParseError> {
+    let (data_section_crc, _) = parse_u32(data)?;
+    Ok(DataEnd { data_section_crc })
+}
+
+/// Parses the content of an MCAP record from a buffer, without copying any string
+/// or array fields. use the [`lifetime::IntoStatic`] implementation to create
+/// an owned copy of the resulting parsed Record.
+pub fn parse_record(opcode: OpCode, data: &[u8]) -> Result<Record<'_>, ParseError> {
+    match opcode {
+        OpCode::Header => Ok(Record::Header(parse_header(data)?)),
+        OpCode::Footer => Ok(Record::Footer(parse_footer(data)?)),
+        OpCode::Schema => Ok(Record::Schema(parse_schema(data)?)),
+        OpCode::Channel => Ok(Record::Channel(parse_channel(data)?)),
+        OpCode::Message => Ok(Record::Message(parse_message(data)?)),
+        OpCode::Chunk => Ok(Record::Chunk(parse_chunk(data)?)),
+        OpCode::MessageIndex => Ok(Record::MessageIndex(parse_message_index(data)?)),
+        OpCode::ChunkIndex => Ok(Record::ChunkIndex(parse_chunk_index(data)?)),
+        OpCode::Attachment => Ok(Record::Attachment(parse_attachment(data)?)),
+        OpCode::AttachmentIndex => Ok(Record::AttachmentIndex(parse_attachment_index(data)?)),
+        OpCode::Statistics => Ok(Record::Statistics(parse_statistics(data)?)),
+        OpCode::Metadata => Ok(Record::Metadata(parse_metadata(data)?)),
+        OpCode::MetadataIndex => Ok(Record::MetadataIndex(parse_metadata_index(data)?)),
+        OpCode::SummaryOffset => Ok(Record::SummaryOffset(parse_summary_offset(data)?)),
+        OpCode::DataEnd => Ok(Record::DataEnd(parse_data_end(data)?)),
+        OpCode::UserOpcode(_) => Err(ParseError::OpCodeNotImplemented(opcode)),
+    }
+}
+
+/// Parses only the first fields of an Attachment record, before the `data` field. This
+/// allows a reader to read and parse only the header of an attachment before deciding whether
+/// to use the rest of the attachment.
+pub fn parse_attachment_header(data: &'_ [u8]) -> Result<AttachmentHeader<'_>, ParseError> {
```

Comment:
does `'_` have a special meaning different from `'a`? I thought I had usually seen these lifetimes given letter names.

Reply:
`'_` is the anonymous lifetime, code for "let lifetime elision figure out what to do with these lifetimes". Using named lifetimes here compiles, I actually only moved to this style when prompted by [clippy](https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes).

Reply:
Is there also some default behavior where you can omit the lifetimes entirely and it'll do the same thing by default?

Reply:
Good point, that was missing here.